contracts/MetaIdentityManager.sol

```diff
@@ -0,0 +1,211 @@
+  function addOwner(address sender, Proxy identity, address newOwner)
+    onlyAuthorized
+    onlyOlderOwner(identity, sender)
+    rateLimited(identity, sender)
+  {
+    owners[identity][newOwner] = now;
+    OwnerAdded(identity, newOwner, sender);
+  }
+
+  /// @dev Allows a recoveryKey to add a new owner with userTimeLock waiting time
+  function addOwnerFromRecovery(address sender, Proxy identity, address newOwner)
+    onlyAuthorized
+    onlyRecovery(identity, sender)
+    rateLimited(identity, sender)
+  {
+    if (owners[identity][newOwner] > 0) throw;
+    owners[identity][newOwner] = now;
+    OwnerAdded(identity, newOwner, sender);
+  }
+
+  /// @dev Allows an owner to remove another owner instantly
+  function removeOwner(address sender, Proxy identity, address owner)
+    onlyAuthorized
+    onlyOlderOwner(identity, sender)
+    rateLimited(identity, sender)
+  {
+    delete owners[identity][owner];
+    OwnerRemoved(identity, owner, sender);
+  }
+
+  /// @dev Allows an owner to change the recoveryKey instantly
+  function changeRecovery(address sender, Proxy identity, address recoveryKey)
+    onlyAuthorized
+    onlyOlderOwner(identity, sender)
+    rateLimited(identity, sender)
+    validRecovery(recoveryKey)
+  {
+    recoveryKeys[identity] = recoveryKey;
+    RecoveryChanged(identity, recoveryKey, sender);
+  }
+
+  /// @dev Allows an owner to being process of transfering proxy to new IdentityManager
+  function initiateMigration(address sender, Proxy identity, address newIdManager)
+    onlyAuthorized
+    onlyOlderOwner(identity, sender)
+  {
+    migrations[identity].startTime = now;
+    migrations[identity].newAddress = newIdManager;
+    MigrationInitiated(identity, newIdManager, sender);
+  }
+
+  /// @dev Allows an owner to cancel the process of transfering proxy to new IdentityManager
+  function cancelMigration(address sender, Proxy identity) onlyAuthorized onlyOwner(identity, sender) {
+    address canceledManager = migrations[identity].newAddress;
+    delete migrations[identity];
+    MigrationCanceled(identity, canceledManager, sender);
+  }
+
+  /// @dev Allows an owner to finalize and completly transfer proxy to new IdentityManager
+  /// Note: before transfering to a new address, make sure this address is "ready to recieve" the proxy.
+  /// Not doing so risks the proxy becoming stuck.
+  function finalizeMigration(address sender, Proxy identity) onlyAuthorized onlyOlderOwner(identity, sender) {
+    if (migrations[identity].startTime > 0 && migrations[identity].startTime + adminTimeLock < now) {
+      address newIdManager = migrations[identity].newAddress;
+      delete migrations[identity];
+      identity.transfer(newIdManager);
+      MigrationFinalized(identity, newIdManager, sender);
+    }
+  }
+
+  //Checks that address a is the first input in msg.data.
+  //Has very minimal gas overhead.
+  function checkMessageData(address a) constant internal returns (bool t) {
```

Comment:
Good point - I didn't think of this. Do you think it might make sense to keep in both (as it's not a ton of duplicated code), and there is also the disadvantage that a library adds very small amount of overhead to every call ~1000 gas, so I think it would be worth it to keep in the long term.